src/main/java/com/squareup/auto/value/redacted/AutoValueRedactedExtension.java

```diff
@@ -0,0 +1,129 @@
+
+import com.google.auto.service.AutoService;
+import com.google.auto.value.AutoValueExtension;
+import com.google.common.collect.Lists;
+import com.squareup.javapoet.ArrayTypeName;
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.JavaFile;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.ParameterSpec;
+import com.squareup.javapoet.TypeName;
+import com.squareup.javapoet.TypeSpec;
+import com.squareup.javapoet.TypeVariableName;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.Name;
+
+@AutoService(AutoValueExtension.class)
+public class AutoValueRedactedExtension implements AutoValueExtension {
+
+  @Override
+  public boolean applicable(Context context) {
+    Map<String, ExecutableElement> properties = context.properties();
+    for (ExecutableElement element : properties.values()) {
+      if (isRedacted(element)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  @Override
+  public boolean mustBeAtEnd(Context context) {
+    return false;
+  }
+
+  @Override
+  public String generateClass(Context context, String className, String classToExtend,
+      boolean isFinal) {
+    Name superName = context.autoValueClass().getSimpleName();
+    Map<String, ExecutableElement> properties = context.properties();
+
+    TypeSpec subclass = TypeSpec.classBuilder(className) //
+        .addModifiers(isFinal ? Modifier.FINAL : Modifier.ABSTRACT) //
+        .superclass(TypeVariableName.get(classToExtend)) //
+        .addMethod(generateConstructor(properties)) //
+        .addMethod(generateToString(superName, properties)) //
+        .build();
+
+    JavaFile javaFile = JavaFile.builder(context.packageName(), subclass).build();
+    return javaFile.toString();
+  }
+
+  private MethodSpec generateConstructor(Map<String, ExecutableElement> properties) {
+    List<ParameterSpec> params = Lists.newArrayList();
+    for (Map.Entry<String, ExecutableElement> entry : properties.entrySet()) {
+      TypeName typeName = TypeName.get(entry.getValue().getReturnType());
+      params.add(ParameterSpec.builder(typeName, entry.getKey()).build());
+    }
+
+    StringBuilder body = new StringBuilder("super(");
+    for (int i = properties.size(); i > 0; i--) {
+      body.append("$N");
+      if (i > 1) body.append(", ");
+    }
+    body.append(")");
+
+    return MethodSpec.constructorBuilder() //
+        .addParameters(params) //
+        .addStatement(body.toString(), properties.keySet().toArray()) //
+        .build();
+  }
+
+  private MethodSpec generateToString(Name superName, Map<String, ExecutableElement> properties) {
+    MethodSpec.Builder builder = MethodSpec.methodBuilder("toString") //
+        .addAnnotation(Override.class) //
+        .addModifiers(Modifier.PUBLIC) //
```

Comment:
Hmm... +`final` I think. We don't want another extension erroneously overriding and emitting a non-redacted `toString()`.